feat: doc tool generate provides `referenced by` information on schema

shruti2522 opened the pull request:
<!-- Thank you for contributing to KCL!

Note: 

1. With pull requests:

    - Open your pull request against "main"
    - Your pull request should have no more than two commits, if not you should squash them.
    - It should pass all tests in the available continuous integration systems such as GitHub Actions.
    - You should add/modify tests to cover your proposed code changes.
    - If your pull request contains a new feature, please document it on the README.

2. Please create an issue first to describe the problem.

    We recommend that link the issue with the PR in the following question.
    For more info, check https://kcl-lang.io/docs/community/contribute/
-->

#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 
fix #165 

<!-- You can add issue references here. 
    e.g. 
    fix #123, re #123, 
    fix https://github.com/XXX/issues/44
-->

#### 2. What is the scope of this PR (e.g. component or file name):

<!-- You can add the scope of this change here. 
    e.g. 
    /src/server/core.rs,
    kcl-lang/kcl/src/parser 
-->
kcl-lang/kcl-go/pkg/tools/gen
#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

<!-- You can choose a brief description here -->
- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

<!-- You can add more details here.
    e.g. 
    Call method "XXXX" to ..... in order to ....,
    More details: https://XXXX.com/doc......
-->

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [x] Y 

<!-- You can add more details here.
    e.g. 
    Calling method "XXXX" will cause the "XXXX", "XXXX" modules to be affected.
    More details: https://XXXX.com/doc......
-->

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

<!-- You can choose a brief description here -->
- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [x] Other

<!-- You can add more details here.
e.g. 
The test case in XXXX is used to .....
test cases in /src/tests/XXXXX
test cases https://github.com/XXX/pull/44
benchmark stats: time XXX ms
-->

Peefy commented on pkg/tools/gen/gengo_test.go:
```diff
@@ -49,6 +49,10 @@ schema Company:
 	}
 	goCode := buf.String()
 	fmt.Println(goCode)
+	// Check if the generated Go code includes the ReferencedBy field
```
Why code changed  in the gengo_test?

shruti2522 replied:
I previously implemented this change as part of my approach, but I haven't reverted the code in all files yet. These modifications will be undone once I push the code.

Peefy commented on pkg/tools/gen/genopenapi.go:
```diff
@@ -192,6 +192,7 @@ type KclOpenAPIType struct {
 	Examples             map[string]KclExample      `json:"examples,omitempty"`             // examples
 	ExternalDocs         string                     `json:"externalDocs,omitempty"`         // externalDocs
 	Ref                  string                     `json:"ref,omitempty"`                  // reference to schema path
+	ReferencedBy        []string                    `json:"referencedBy,omitempty"`         // Schemas referencing this schema
```
Run `make fmt` to format all go files.

Peefy commented on pkg/tools/gen/testdata/doc/pkg/md_got/docs/main.md:
```diff
@@ -0,0 +1,75 @@
+## Index
+
+- [Container](#container)
+- [Server](#server) (Referenced by: Container)
```
Why add the `referenced by` info on the TOC?

Peefy commented on pkg/tools/gen/templates/doc/schemaDoc.gotmpl:
```diff
@@ -20,4 +20,4 @@
 {{end}}
 {{end -}}
-{{- end -}}
+{{- end -}}
```
Append a blank line.

Peefy commented on pkg/tools/gen/gendoc.go:
```diff
@@ -211,23 +232,29 @@ func (pkg *KclPackage) getPackageIndexContent(level int, indentation string) str
 func (tpe *KclOpenAPIType) getSchemaIndexContent(level int, indentation string) string {
-	return fmt.Sprintf(`%s- [%s](#%s)
-`, strings.Repeat(indentation, level), tpe.KclExtensions.XKclModelType.Type, strings.ToLower(tpe.KclExtensions.XKclModelType.Type))
+    var referencedBy strings.Builder
```
Why add the `referenced by` info on the TOC?

shruti2522 replied:
Yes, you are right. We want the TOC to be easy to read and understand. Initially, I thought adding `referenced by` info to the TOC would be helpful because it could provide additional context about where certain schema types are used.

Peefy replied:
A single section named `referenced by` and add quick links for the reference schema maybe useful.

shruti2522 committed:
feat: add refernced by info to doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

modified template to provide referenced by info

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

defined refernced by in KclOpenAPIType struct

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated document generation

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

corrected formatting

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated gpyrpc.proto

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

feat: add refernced by info to doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated GetFullSchemaType api

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

corrected formatting

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated template

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated gpyrpc.proto

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated base schema type

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated base-schema type

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated api.go

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated getKclOpenAPIType to provide referenced by info

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated gendoc.go

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated template

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated main.md

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

feat: add refernced by info to doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated GenKclOpenAPIType

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

removed getBaseSchemaRecursive

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

added unit tests for referencedBy in

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

added unit tests for referencedBy in genopenapi.go

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated doc template

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

reverted changes in pkg/md/main/md

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

removed extra whitespaces in doc template

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

updated doc template

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

feat: add referenced by info to doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

feat: add referenced by info to doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

Peefy commented on pkg/tools/gen/testdata/doc/pkg/md/main.md:
```diff
@@ -49,6 +49,8 @@ Server is the common user interface for long-running services adopting the best
 |**union** `required`|"abc" \| 123 \| True \| 1.11 \| [Container](#container) \| units.NumberMultiplier \| 1M|||
 |**union2** `required`|"abc" \| "def"||"abc"|
 |**workloadType** `required`|str|Use this attribute to specify which kind of long-running service you want.<br />Valid values: Deployment, CafeDeployment.<br />See also: kusion_models/core/v1/workload_metadata.k.|"Deployment"|
+#### Referenced By
```
CI failed.

Peefy commented on pkg/spec/gpyrpc/gpyrpc.proto:
```diff
@@ -479,6 +479,7 @@ message KclType {
 	string pkg_path = 13;                // `pkg_path` represents the path name of the package where the attribute is located.
 	string description = 14;             // `description` represents the document of the attribute.
 	map<string, Example> examples = 15;  // A map object to hold examples, the key is the example name.
+	map<string, KclType> referencedBy = 16;          // contains referenced by info for schema
```
Why add referenceBy here? It should be same as the spec: https://github.com/kcl-lang/kcl/blob/main/kclvm/spec/gpyrpc/gpyrpc.proto#L482

Peefy commented on pkg/tools/gen/testdata/doc/pkg/md_got/docs/main.md:
```diff
@@ -0,0 +1,80 @@
+#### Referenced By
+
+
+- {object   [] false Container is the common user interface for long-running services. map[image:0xc0007964e0 name:0xc000796680] [name image] <nil> <nil> map[]   map[] 0xc00047a700}
```
Error Format

Peefy commented on pkg/tools/gen/templates/doc/schemaDoc.gotmpl:
```diff
@@ -11,13 +11,18 @@
 {{range $name, $property := $Data.Properties}}|**{{$name}}**{{if containsString $Data.Required $name }} `required`{{end}}{{if $property.ReadOnly}} `readOnly`{{end}}|{{kclType $property $EscapeHtml}}|{{if ne $property.Description ""}}{{escapeHtml $property.Description $EscapeHtml}}{{end}}|{{escapeHtml $property.Default $EscapeHtml}}|
-{{end}}{{if ne (len $Data.Examples) 0}}#### Examples
+{{end}}{{if ne (len $Data.ReferencedBy) 0}}#### Referenced By
+
```
Remove blank lines.

Peefy commented on pkg/tools/gen/templates/doc/schemaDoc.gotmpl:
```diff
@@ -11,13 +11,18 @@
 | name | type | description | default value |
 | --- | --- | --- | --- |
 {{range $name, $property := $Data.Properties}}|**{{$name}}**{{if containsString $Data.Required $name }} `required`{{end}}{{if $property.ReadOnly}} `readOnly`{{end}}|{{kclType $property $EscapeHtml}}|{{if ne $property.Description ""}}{{escapeHtml $property.Description $EscapeHtml}}{{end}}|{{escapeHtml $property.Default $EscapeHtml}}|
-{{end}}{{if ne (len $Data.Examples) 0}}#### Examples
```
Add a blank line here.

Peefy commented on pkg/tools/gen/templates/doc/schemaDoc.gotmpl:
```diff
@@ -19,5 +19,11 @@
+{{end}}
 {{end -}}
-{{- end -}}
+{{- end -}}
```
Append a blank here.

Peefy commented on pkg/tools/gen/testdata/doc/pkg/md_got/docs/main.md:
```diff
@@ -0,0 +1,76 @@
+# main
```
md_got/docs/main.md is the expect test result for the md/docs/main.md, you can copy the content of xmd_got/docs/main.md to md/docs/main.md and delete md_got/docs/main.md

Peefy commented on pkg/tools/gen/testdata/doc/pkg/md_got/docs/main.md:
```diff
@@ -0,0 +1,76 @@
+| --- | --- | --- | --- |
+|**image** `required`|str|||
+|**name** `required`|str|The name of the long-running container.||
+### Server
```
There need to be some blank lines here, please check the modifications in pkg/tools/gen/templates/doc/schemaDoc.gotmpl

Peefy commented on pkg/tools/gen/genopenapi.go:
```diff
@@ -375,10 +377,15 @@ func GetPkgDir(base string, pkgName string) string {
 
 // GetKclOpenAPIType converts the kcl.KclType(the representation of Type in KCL API) to KclOpenAPIType(the representation of Type in KCL Open API)
-func GetKclOpenAPIType(pkgPath string, from *kcl.KclType, nested bool) *KclOpenAPIType {
+func GetKclOpenAPIType(pkgPath string, from *kcl.KclType, nested bool, base_schema *gpyrpc.KclType) *KclOpenAPIType {
```
I think the parameter `base_schema` is unnecessary here, because base_schema is existed in `from.BaseSchema`

Peefy commented on pkg/kcl/api.go:
```diff
@@ -408,7 +408,31 @@ func GetFullSchemaType(pathList []string, schemaName string, opts ...Option) ([]
+	result := []*gpyrpc.KclType{}
+	for _, kclType := range resp.SchemaTypeList {
+		if kclType.BaseSchema != nil {
+			baseSchema, err := getBaseSchemaRecursive(kclType.BaseSchema)
```
I think it is not necessary to recursively obtain the base schema here because they are not directly dependent on each other.

Peefy commented on pkg/tools/gen/gendoc.go:
```diff
@@ -159,8 +159,41 @@ func addOrCreateSchema(pkg *KclPackage, schemaName string, schema *KclOpenAPITyp
 		pkg.schemaMapping = map[string]*KclOpenAPIType{schemaName: schema}
 	} else {
-		pkg.schemaMapping[schemaName] = schema
+		if existingSchema, ok := pkg.schemaMapping[schemaName]; ok {
```
Cloud you please add more unit test cases for the referenced by feature?

Peefy commented on pkg/tools/gen/genopenapi.go:
```diff
@@ -376,9 +377,14 @@ func GetPkgDir(base string, pkgName string) string {
 // GetKclOpenAPIType converts the kcl.KclType(the representation of Type in KCL API) to KclOpenAPIType(the representation of Type in KCL Open API)
 func GetKclOpenAPIType(pkgPath string, from *kcl.KclType, nested bool) *KclOpenAPIType {
+	var referencedBy []string
+	if from.BaseSchema != nil {
```
I think one schema is refernecd by its child schema instead of the schema.

shruti2522 committed:
updated referencing logic

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
update referncedBy logic

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
fmt check

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
add tests for referencedBy

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

Peefy commented on pkg/spec/gpyrpc/gpyrpc.proto:
```diff
@@ -481,7 +481,7 @@ message KclType {
 	string description = 14;             // `description` represents the document of the attribute.
 	map<string, Example> examples = 15;  // A map object to hold examples, the key is the example name.
-	KclType base_schema = 16;            // contains referenced by info for schema
+	KclType base_schema = 16;          // contains referenced by info for schema
```
Prevent the code change.

shruti2522 committed:
prevent code change in gpyrpc.proto

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

Peefy commented on pkg/tools/gen/genopenapi.go:
```diff
@@ -376,9 +377,17 @@ func GetPkgDir(base string, pkgName string) string {
+	var referencedBy []string
+	schemaID := SchemaId(pkgPath, from)
+	if from.BaseSchema != nil {
+		childSchema := GetKclOpenAPIType(pkgPath, from.BaseSchema, true)
```
Why `childSchema := BaseScheama`

shruti2522 committed:
base schema reference

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
fmt check

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
fmt check

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

Peefy commented on pkg/tools/gen/genopenapi_test.go:
```diff
@@ -24,8 +24,7 @@ func TestExportOpenAPIV3Spec(t *testing.T) {
 
-	expect := `{"default":"","description":"AppConfiguration is a developer-centric definition that describes how to run an Application. This application model builds upon a decade of experience at AntGroup running super large scale internal developer platform, combined with best-of-breed ideas and practices from the community.","example":{"Default example":{"value":"# Instantiate an App with a long ...
-
+	expect := `{"default":"","description":"AppConfiguration is a developer-centric definition that describes how to run an Application. This application model builds upon a decade of experience at AntGroup running super large scale internal developer platform, combined with best-of-breed ideas and practices from the community.","example":{"Default example":{"value":"# Instantiate an App with a long ...
```
WorkloadBase referencedby Service and Job not Service referencedby WorkloadBase.

shruti2522 committed:
revert old tests

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

shruti2522 committed:
baseSchema info on doc gen

Signed-off-by: shruti2522 <shruti.apc01@gmail.com>

Peefy reviewed (approved):
LGTM
